<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Munging summary statistics · GeneticsMakie.jl</title><meta name="title" content="Munging summary statistics · GeneticsMakie.jl"/><meta property="og:title" content="Munging summary statistics · GeneticsMakie.jl"/><meta property="twitter:title" content="Munging summary statistics · GeneticsMakie.jl"/><meta name="description" content="Documentation for GeneticsMakie.jl."/><meta property="og:description" content="Documentation for GeneticsMakie.jl."/><meta property="twitter:description" content="Documentation for GeneticsMakie.jl."/><meta property="og:url" content="https://mmkim1210.github.io/GeneticsMakie.jl/examples/summary/"/><meta property="twitter:url" content="https://mmkim1210.github.io/GeneticsMakie.jl/examples/summary/"/><link rel="canonical" href="https://mmkim1210.github.io/GeneticsMakie.jl/examples/summary/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.svg" alt="GeneticsMakie.jl logo"/></a><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../gtf/">Parsing GENCODE</a></li><li class="is-active"><a class="tocitem" href>Munging summary statistics</a></li><li><a class="tocitem" href="../genes/">Plotting genes</a></li><li><a class="tocitem" href="../isoforms/">Plotting isoforms</a></li><li><a class="tocitem" href="../locus/">Plotting LocusZoom</a></li><li><a class="tocitem" href="../gwas/">Plotting GWAS</a></li><li><a class="tocitem" href="../twas/">Plotting TWAS</a></li><li><a class="tocitem" href="../loops/">Plotting loops</a></li><li><a class="tocitem" href="../peaks/">Plotting peaks</a></li></ul></li><li><a class="tocitem" href="../../api/">API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Examples</a></li><li class="is-active"><a href>Munging summary statistics</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Munging summary statistics</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/master/docs/src/examples/summary.md#" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Munging-summary-statistics"><a class="docs-heading-anchor" href="#Munging-summary-statistics">Munging summary statistics</a><a id="Munging-summary-statistics-1"></a><a class="docs-heading-anchor-permalink" href="#Munging-summary-statistics" title="Permalink"></a></h1><p>GWAS summary statistics come in a variety of shapes and flavors, so harmonizing them is crucial in making our lives easier when trying to visualize their results. We can  take a peak at GWAS results for height and weight, the two classic anthropometric traits. </p><pre><code class="language-julia hljs">using Pkg
Pkg.add([&quot;GeneticsMakie&quot;, &quot;CSV&quot;, &quot;DataFrames&quot;, &quot;Arrow&quot;])</code></pre><pre><code class="language-julia hljs">using GeneticsMakie, CSV, DataFrames, Arrow, Downloads
gwas = Dict(
    &quot;height&quot; =&gt; &quot;https://portals.broadinstitute.org/collaboration/giant/images/6/63/Meta-analysis_Wood_et_al%2BUKBiobank_2018.txt.gz&quot;,
    &quot;weight&quot; =&gt; &quot;https://portals.broadinstitute.org/collaboration/giant/images/c/c8/Meta-analysis_Locke_et_al%2BUKBiobank_2018_UPDATED.txt.gz&quot;
)
isdir(&quot;data/gwas&quot;) || mkdir(&quot;data/gwas&quot;)
dfs = DataFrame[]
for key in keys(gwas)
    url = gwas[key]
    isfile(&quot;data/gwas/$(basename(url))&quot;) || Downloads.download(url, &quot;data/gwas/$(basename(url))&quot;)
    push!(dfs, CSV.read(&quot;data/gwas/$(basename(url))&quot;, DataFrame; comment = &quot;##&quot;, missingstring = [&quot;NA&quot;]))
end</code></pre><p>To harmonize summary statistics, we run a single command <code>GeneticsMakie.mungesumstats!</code>.  For all downstream plotting functions, we note that GWAS summary statistics should be pre-harmonized.</p><pre><code class="language-julia hljs">GeneticsMakie.mungesumstats!(dfs)</code></pre><p>To find GWAS loci for each phenotype that are separated by at least 1 Mb,  we can execute the following command.</p><pre><code class="language-julia hljs">GeneticsMakie.findgwasloci(dfs[1])
GeneticsMakie.findgwasloci(dfs[2])</code></pre><p>To find every GWAS loci across multiple phenotypes that are separated by at least 1 Mb, we can instead run the following command.</p><pre><code class="language-julia hljs">GeneticsMakie.findgwasloci(dfs)</code></pre><p>Such an exhaustive list of loci can then be iterated through and visualized by  <a href="../locus/#Plotting-LocusZooom">Plotting LocusZooom</a>.</p><p>We can also find the closest gene to each index SNP in GWAS loci after <a href="../gtf/#Parsing-GENCODE">Parsing GENCODE</a>.</p><pre><code class="language-julia hljs">file = &quot;gencode.v39lift37.annotation.gtf.arrow&quot;
gencode = Arrow.Table(&quot;data/gencode/$(file)&quot;)|&gt; DataFrame

loci = GeneticsMakie.findgwasloci(dfs[1])
GeneticsMakie.findclosestgene(loci, gencode)
GeneticsMakie.findclosestgene(loci, gencode; start = true) # closest gene from gene start site
GeneticsMakie.findclosestgene(loci, gencode; proteincoding = true) # closest &quot;protein-coding&quot; gene</code></pre><p>To reduce memory intake, we can store and load GWAS summary statistics as Arrow files. </p><pre><code class="language-julia hljs">for (i, key) in enumerate(keys(gwas))
    Arrow.write(&quot;data/gwas/$(key).arrow&quot;, dfs[i])
end</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../gtf/">« Parsing GENCODE</a><a class="docs-footer-nextpage" href="../genes/">Plotting genes »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="auto">Automatic (OS)</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.1.2 on <span class="colophon-date" title="Saturday 28 October 2023 22:55">Saturday 28 October 2023</span>. Using Julia version 1.9.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
