<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Plotting GWAS · GeneticsMakie.jl</title><meta name="title" content="Plotting GWAS · GeneticsMakie.jl"/><meta property="og:title" content="Plotting GWAS · GeneticsMakie.jl"/><meta property="twitter:title" content="Plotting GWAS · GeneticsMakie.jl"/><meta name="description" content="Documentation for GeneticsMakie.jl."/><meta property="og:description" content="Documentation for GeneticsMakie.jl."/><meta property="twitter:description" content="Documentation for GeneticsMakie.jl."/><meta property="og:url" content="https://mmkim1210.github.io/GeneticsMakie.jl/examples/gwas/"/><meta property="twitter:url" content="https://mmkim1210.github.io/GeneticsMakie.jl/examples/gwas/"/><link rel="canonical" href="https://mmkim1210.github.io/GeneticsMakie.jl/examples/gwas/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.svg" alt="GeneticsMakie.jl logo"/></a><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../gtf/">Parsing GENCODE</a></li><li><a class="tocitem" href="../summary/">Munging summary statistics</a></li><li><a class="tocitem" href="../genes/">Plotting genes</a></li><li><a class="tocitem" href="../isoforms/">Plotting isoforms</a></li><li><a class="tocitem" href="../locus/">Plotting LocusZoom</a></li><li class="is-active"><a class="tocitem" href>Plotting GWAS</a></li><li><a class="tocitem" href="../twas/">Plotting TWAS</a></li><li><a class="tocitem" href="../loops/">Plotting loops</a></li><li><a class="tocitem" href="../peaks/">Plotting peaks</a></li></ul></li><li><a class="tocitem" href="../../api/">API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Examples</a></li><li class="is-active"><a href>Plotting GWAS</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Plotting GWAS</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/master/docs/src/examples/gwas.md#" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Plotting-GWAS"><a class="docs-heading-anchor" href="#Plotting-GWAS">Plotting GWAS</a><a id="Plotting-GWAS-1"></a><a class="docs-heading-anchor-permalink" href="#Plotting-GWAS" title="Permalink"></a></h1><p>After <a href="../summary/#Munging-summary-statistics">Munging summary statistics</a>, we can use <code>GeneticsMakie.plotgwas!</code>  to draw Manhattan plots.</p><pre><code class="language-julia hljs">using Pkg
Pkg.add([&quot;GeneticsMakie&quot;, &quot;CairoMakie&quot;, &quot;DataFrames&quot;, &quot;Arrow&quot;])</code></pre><pre><code class="language-julia hljs">using GeneticsMakie, CairoMakie, DataFrames, Arrow
dfs = DataFrame[]
for key in [&quot;height&quot;, &quot;weight&quot;]
    push!(dfs, Arrow.Table(&quot;data/gwas/$(key).arrow&quot;)|&gt; DataFrame)
end
titles = [&quot;Height (Yengo et al. 2018)&quot;, &quot;Weight (Yengo et al. 2018)&quot;]</code></pre><pre><code class="language-julia hljs">f = Figure(resolution = (408, 792))
axs = [Axis(f[i, 1]) for i in 1:length(titles)]
for i in eachindex(titles)
    GeneticsMakie.plotgwas!(axs[i], dfs[i])
    hidespines!(axs[i], :t, :r)
    Label(f[i, 1, Top()], text = &quot;$(titles[i])&quot;, fontsize = 8)
    rowsize!(f.layout, i, 50)
    i == length(titles) ? axs[i].xlabel = &quot;Chromosome&quot; : axs[i].xlabel = &quot;&quot;
end
rowgap!(f.layout, 10)
resize_to_layout!(f)
f</code></pre><p><img src="../../figs/manhattan.png" alt/></p><p>By default, <code>GeneticsMakie.plotgwas!</code> highlights the genome-wide significant threshold and corresponding significant variants. We can turn off this option by using the <code>linecolor</code> and <code>scattercolor</code> keyword arguments. </p><pre><code class="language-julia hljs">f = Figure(resolution = (408, 792))
axs = [Axis(f[i, 1]) for i in 1:length(titles)]
for i in eachindex(titles)
    GeneticsMakie.plotgwas!(axs[i], dfs[i]; linecolor = nothing, scattercolor = nothing)
    hidespines!(axs[i], :t, :r)
    Label(f[i, 1, Top()], text = &quot;$(titles[i])&quot;, fontsize = 8)
    rowsize!(f.layout, i, 50)
    i == length(titles) ? axs[i].xlabel = &quot;Chromosome&quot; : axs[i].xlabel = &quot;&quot;
end
rowgap!(f.layout, 10)
resize_to_layout!(f)
f</code></pre><p><img src="../../figs/manhattan-nocolor.png" alt/></p><p>We can color even and odd chromosomes with different colors by using the <code>chromcolors</code> keyword argument. </p><pre><code class="language-julia hljs">f = Figure(resolution = (408, 792))
axs = [Axis(f[i, 1]) for i in 1:length(titles)]
for i in eachindex(titles)
    GeneticsMakie.plotgwas!(axs[i], dfs[i]; linecolor = nothing, scattercolor = nothing, 
        chromcolors = [&quot;#389826&quot;, &quot;#9658B2&quot;])
    hidespines!(axs[i], :t, :r)
    Label(f[i, 1, Top()], text = &quot;$(titles[i])&quot;, fontsize = 8)
    rowsize!(f.layout, i, 50)
    i == length(titles) ? axs[i].xlabel = &quot;Chromosome&quot; : axs[i].xlabel = &quot;&quot;
end
rowgap!(f.layout, 10)
resize_to_layout!(f)
f</code></pre><p><img src="../../figs/manhattan-chromosomes.png" alt/></p><p>We can then use <code>GeneticsMakie.plotqq!</code> to draw QQ plots.</p><pre><code class="language-julia hljs">f = Figure(resolution = (408, 792))
axs = [Axis(f[2, i]) for i in 1:length(titles)]
for i in eachindex(titles)
    GeneticsMakie.plotqq!(axs[i], dfs[i]; ystep = 5)
    axs[i].xlabel = &quot;&quot;
    axs[i].ylabel = &quot;&quot;
    ylims!(axs[i], 0, 40)
    i &gt; 1 ? hideydecorations!(axs[i]) : nothing
end
for (i, title) in enumerate(titles)
    Box(f[1, i], color = :gray90)
    Label(f[1, i], title, tellwidth = false, fontsize = 8, padding = (0, 0, 3, 3))
end
Label(f[3, 1:length(titles)], text = &quot;Expected -log[p]&quot;, fontsize = 8)
Label(f[2, 0], text = &quot;Observed -log[p]&quot;, fontsize = 8, rotation = pi / 2, tellheight = false)
rowsize!(f.layout, 2, Aspect(2, 1))
colgap!(f.layout, 5)
rowgap!(f.layout, 1, 0)
rowgap!(f.layout, 2, 5)
resize_to_layout!(f)
f</code></pre><p><img src="../../figs/qq.png" alt/></p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../locus/">« Plotting LocusZoom</a><a class="docs-footer-nextpage" href="../twas/">Plotting TWAS »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="auto">Automatic (OS)</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.1.2 on <span class="colophon-date" title="Saturday 28 October 2023 23:12">Saturday 28 October 2023</span>. Using Julia version 1.9.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
