var documenterSearchIndex = {"docs":
[{"location":"","page":"Home","title":"Home","text":"CurrentModule = GeneticsMakie","category":"page"},{"location":"#GeneticsMakie","page":"Home","title":"GeneticsMakie","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Documentation for GeneticsMakie.","category":"page"},{"location":"","page":"Home","title":"Home","text":"","category":"page"},{"location":"","page":"Home","title":"Home","text":"Modules = [GeneticsMakie]","category":"page"},{"location":"#GeneticsMakie.coordinategenes-Tuple{AbstractString, Real, Real, DataFrames.DataFrame, Real}","page":"Home","title":"GeneticsMakie.coordinategenes","text":"coordinategenes(chromosome::AbstractString, range1::Real, range2::Real, gencode::DataFrame, height::Real)\n\nSubset gencode to a given chromosome and genomic range between range1 and range2, and determine coordinates of exons for each gene residing in the genomic region.\n\n\n\n\n\n","category":"method"},{"location":"#GeneticsMakie.coordinategwas-Tuple{Vector{DataFrames.DataFrame}}","page":"Home","title":"GeneticsMakie.coordinategwas","text":"coordinategwas(gwas::Vector{DataFrame}; ymax, chr, bp, p, freey)\n\nDetermine shared coordinates for a set of gwas.\n\n\n\n\n\n","category":"method"},{"location":"#GeneticsMakie.coordinateisforms-Tuple{AbstractString, DataFrames.DataFrame, Union{Nothing, AbstractVector{<:AbstractString}}, Real, Bool, Symbol}","page":"Home","title":"GeneticsMakie.coordinateisforms","text":"coordinateisforms(gene::AbstractString, gencode::DataFrame, orderby::Union{Nothing, AbstractVector{<:AbstractString}}, height::Real, text::Bool, textpos::Symbol)\n\nSubset gencode to a gene and determine coordinates of exons for each isoform.\n\n\n\n\n\n","category":"method"},{"location":"#GeneticsMakie.findgene-Tuple{AbstractString, DataFrames.DataFrame}","page":"Home","title":"GeneticsMakie.findgene","text":"findgene(gene::AbstractString, gencode::DataFrame)\n\nFind chromosome, start, and stop sites for the gene of interest.\n\n\n\n\n\n","category":"method"},{"location":"#GeneticsMakie.labelgenome-Tuple{GridLayoutBase.GridPosition, AbstractString, Real, Real}","page":"Home","title":"GeneticsMakie.labelgenome","text":"labelgenome(g::GridPosition, chromosome::AbstractString, range1::Real, range2::Real)\n\n\n\n\n\n","category":"method"},{"location":"#GeneticsMakie.parsegtf!-Tuple{DataFrames.DataFrame}","page":"Home","title":"GeneticsMakie.parsegtf!","text":"parsegtf!(gencode::DataFrame)\n\nParse gencode by extracting gene_id, gene_name, gene_type, transcript_id, transcript_support_level information from the info column.\n\n\n\n\n\n","category":"method"},{"location":"#GeneticsMakie.plotgenes!-Tuple{Makie.MakieLayout.Axis, AbstractString, DataFrames.DataFrame}","page":"Home","title":"GeneticsMakie.plotgenes!","text":"plotgenes!(ax::Axis, gene::AbstractString, gencode::DataFrame; window)\n\nPlot collapsed gene bodies for genes within a certain window around gene.  The default window is 1 Mb.\n\n\n\n\n\n","category":"method"},{"location":"#GeneticsMakie.plotgenes!-Tuple{Makie.MakieLayout.Axis, AbstractString, Real, DataFrames.DataFrame}","page":"Home","title":"GeneticsMakie.plotgenes!","text":"plotgenes!(ax::Axis, chromosome::AbstractString, bp::Real, gencode::DataFrame; window)\n\nPlot collapsed gene bodies for genes within a given chromosome and a certain window around a genomic coordinate bp. The default window is 1 Mb.\n\n\n\n\n\n","category":"method"},{"location":"#GeneticsMakie.plotgenes!-Tuple{Makie.MakieLayout.Axis, AbstractString, Real, Real, DataFrames.DataFrame}","page":"Home","title":"GeneticsMakie.plotgenes!","text":"plotgenes!(ax::Axis, chromosome::AbstractString, range1::Real, range2::Real, gencode::DataFrame; height, exoncolor, textcolor, highlight)\n\nPlot collapsed gene bodies for genes within a given chromosome and genomic range  between range1 and range2. Optionally, height of exons can be adjusted using height and color of exons or gene names adjusted using exoncolor and textcolor, respectively. Certain genes can be highlighted via highlight.\n\n\n\n\n\n","category":"method"},{"location":"#GeneticsMakie.plotgwas!-Tuple{Makie.MakieLayout.Axis, DataFrames.DataFrame, Int64, Real, Real, DataFrames.DataFrame}","page":"Home","title":"GeneticsMakie.plotgwas!","text":"plotgwas!(ax::Axis, df::DataFrame, i::Int, ymax::Real, xmax::Real, ticks::DataFrame; xlabel, ylabel, ystep, chr, p, sigline, sigcolor)\n\nPlot gwas results with coordinates from coordinategwas, namely df for phenotype i with x and y limits, xmax and ymax, respectively. The position of x ticks are determined by ticks.\n\n\n\n\n\n","category":"method"},{"location":"#GeneticsMakie.plotisoforms!-Tuple{Makie.MakieLayout.Axis, AbstractString, DataFrames.DataFrame}","page":"Home","title":"GeneticsMakie.plotisoforms!","text":"plotisoforms!(ax::Axis, gene::AbstractString, gencode::DataFrame; orderby, height, text, textpos)\n\nPlot each isoform of a given gene in a separate row. Optionally, order of isoforms can be changed by orderby, height of exons can be adjusted using height, and color of exons or gene names adjusted using exoncolor and textcolor, respectively. The presence or absence of text label of isoforms can be specified using text with its position textpos.\n\n\n\n\n\n","category":"method"},{"location":"#GeneticsMakie.plotld!-Tuple{Makie.MakieLayout.Axis, AbstractMatrix}","page":"Home","title":"GeneticsMakie.plotld!","text":"plotld!(ax::Axis, LD::AbstractMatrix; color)\n\nVisualize a correlation matrix LD with the diagonal elements on the x-axis.\n\n\n\n\n\n","category":"method"},{"location":"#GeneticsMakie.plotlocus!-Tuple{Makie.MakieLayout.Axis, AbstractString, DataFrames.DataFrame, DataFrames.DataFrame}","page":"Home","title":"GeneticsMakie.plotlocus!","text":"plotlocus!(ax::Axis, gene::AbstractString, gwas::DataFrame, gencode::DataFrame; window, colorld, ref, snp, ymax, title)\n\nPlot gwas results within a certain window around gene. The default window is 1 Mb.\n\n\n\n\n\n","category":"method"},{"location":"#GeneticsMakie.plotlocus!-Tuple{Makie.MakieLayout.Axis, AbstractString, Real, DataFrames.DataFrame}","page":"Home","title":"GeneticsMakie.plotlocus!","text":"plotlocus!(ax::Axis, chromosome::AbstractString, bp::Real, gwas::DataFrame; window, colorld, ref, snp, ymax, title)\n\nPlot gwas results within a given chromosome and a certain window around a  genomic coordinate bp. The default window is 1 Mb.\n\n\n\n\n\n","category":"method"},{"location":"#GeneticsMakie.plotlocus!-Tuple{Makie.MakieLayout.Axis, AbstractString, Real, Real, DataFrames.DataFrame}","page":"Home","title":"GeneticsMakie.plotlocus!","text":"plotlocus!(ax::Axis, chromosome::AbstractString, range1::Real, range2::Real, gwas::DataFrame; colorld, ref, snp, ymax)\n\nPlot gwas results within a given chromosome and genomic range between range1  and range2. Optionally, SNPs can be colored by LD via colorld using ref. The default SNP for which LD is calculated is index SNP, which can be changed to snp.\n\n\n\n\n\n","category":"method"},{"location":"#GeneticsMakie.plotqq!-Tuple{Makie.MakieLayout.Axis, AbstractVector}","page":"Home","title":"GeneticsMakie.plotqq!","text":"plotqq!(ax::Axis, p::AbstractVector; title, xlabel, ylabel, xstep, ystep)\n\nPlot QQ plot of p values where the expected distribution is the uniform distribution. \n\n\n\n\n\n","category":"method"}]
}
