<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API · GeneticsMakie</title><meta name="title" content="API · GeneticsMakie"/><meta property="og:title" content="API · GeneticsMakie"/><meta property="twitter:title" content="API · GeneticsMakie"/><meta name="description" content="Documentation for GeneticsMakie."/><meta property="og:description" content="Documentation for GeneticsMakie."/><meta property="twitter:description" content="Documentation for GeneticsMakie."/><meta property="og:url" content="https://mmkim1210.github.io/GeneticsMakie.jl/api/"/><meta property="twitter:url" content="https://mmkim1210.github.io/GeneticsMakie.jl/api/"/><link rel="canonical" href="https://mmkim1210.github.io/GeneticsMakie.jl/api/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.svg" alt="GeneticsMakie logo"/></a><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../examples/gtf/">Parsing GENCODE</a></li><li><a class="tocitem" href="../examples/summary/">Munging summary statistics</a></li><li><a class="tocitem" href="../examples/genes/">Plotting genes</a></li><li><a class="tocitem" href="../examples/isoforms/">Plotting isoforms</a></li><li><a class="tocitem" href="../examples/locus/">Plotting LocusZoom</a></li><li><a class="tocitem" href="../examples/gwas/">Plotting GWAS</a></li><li><a class="tocitem" href="../examples/twas/">Plotting TWAS</a></li><li><a class="tocitem" href="../examples/loops/">Plotting loops</a></li><li><a class="tocitem" href="../examples/peaks/">Plotting peaks</a></li></ul></li><li class="is-active"><a class="tocitem" href>API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>API</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/master/docs/src/api.md#" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="API"><a class="docs-heading-anchor" href="#API">API</a><a id="API-1"></a><a class="docs-heading-anchor-permalink" href="#API" title="Permalink"></a></h1><ul><li><a href="#GeneticsMakie.findclosestgene-Tuple{AbstractString, Real, DataFrames.DataFrame}"><code>GeneticsMakie.findclosestgene</code></a></li><li><a href="#GeneticsMakie.findgene-Tuple{AbstractString, DataFrames.DataFrame}"><code>GeneticsMakie.findgene</code></a></li><li><a href="#GeneticsMakie.findgwasloci-Tuple{DataFrames.DataFrame}"><code>GeneticsMakie.findgwasloci</code></a></li><li><a href="#GeneticsMakie.labelgenome-Tuple{GridLayoutBase.GridPosition, AbstractString, Real, Real}"><code>GeneticsMakie.labelgenome</code></a></li><li><a href="#GeneticsMakie.mungesumstats!-Tuple{Vector{DataFrames.DataFrame}}"><code>GeneticsMakie.mungesumstats!</code></a></li><li><a href="#GeneticsMakie.parsegtf!-Tuple{DataFrames.DataFrame}"><code>GeneticsMakie.parsegtf!</code></a></li><li><a href="#GeneticsMakie.plotgenes!-Tuple{Makie.Axis, AbstractString, Real, Real, DataFrames.DataFrame}"><code>GeneticsMakie.plotgenes!</code></a></li><li><a href="#GeneticsMakie.plotgenes!-Tuple{Makie.Axis, AbstractString, Real, Real, Tuple{AbstractVector, AbstractVector}, DataFrames.DataFrame}"><code>GeneticsMakie.plotgenes!</code></a></li><li><a href="#GeneticsMakie.plotgwas!-Tuple{Makie.Axis, DataFrames.DataFrame}"><code>GeneticsMakie.plotgwas!</code></a></li><li><a href="#GeneticsMakie.plotisoforms!-Tuple{Makie.Axis, AbstractString, DataFrames.DataFrame}"><code>GeneticsMakie.plotisoforms!</code></a></li><li><a href="#GeneticsMakie.plotld-Tuple"><code>GeneticsMakie.plotld</code></a></li><li><a href="#GeneticsMakie.plotlocus!-Tuple{Makie.Axis, AbstractString, Real, Real, DataFrames.DataFrame}"><code>GeneticsMakie.plotlocus!</code></a></li><li><a href="#GeneticsMakie.plotloops!-Tuple{Makie.Axis, AbstractString, Real, Real, DataFrames.AbstractDataFrame}"><code>GeneticsMakie.plotloops!</code></a></li><li><a href="#GeneticsMakie.plotqq!-Tuple{Makie.Axis, AbstractVector}"><code>GeneticsMakie.plotqq!</code></a></li><li><a href="#GeneticsMakie.plotrg-Tuple"><code>GeneticsMakie.plotrg</code></a></li></ul><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneticsMakie.findclosestgene-Tuple{AbstractString, Real, DataFrames.DataFrame}" href="#GeneticsMakie.findclosestgene-Tuple{AbstractString, Real, DataFrames.DataFrame}"><code>GeneticsMakie.findclosestgene</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">findclosestgene(chr::AbstractString, bp::Real, gencode::DataFrame; start::Bool, proteincoding::Bool)
findclosestgene(df::DataFrame, gencode::DataFrame; start::Bool, proteincoding::Bool)</code></pre><p>Find the closest gene(s) to a genomic coordinate or a list of genomic coordinates using <code>gencode</code>. </p><p>Optionally, the closest gene can be defined from the gene start site using <code>start</code>, and only protein coding genes can be considered using <code>proteincoding</code>.  The default <code>start</code> and <code>proteincoding</code> are <code>false</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/d595fc776b2a561097a6e79d1291ca902361d09d/src/mungesumstats.jl#L346-L355">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneticsMakie.findgene-Tuple{AbstractString, DataFrames.DataFrame}" href="#GeneticsMakie.findgene-Tuple{AbstractString, DataFrames.DataFrame}"><code>GeneticsMakie.findgene</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">findgene(gene::AbstractString, gencode::DataFrame)</code></pre><p>Find chromosome, gene start, and gene stop sites for the <code>gene</code> of interest.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/d595fc776b2a561097a6e79d1291ca902361d09d/src/parsegtf.jl#L34-L38">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneticsMakie.findgwasloci-Tuple{DataFrames.DataFrame}" href="#GeneticsMakie.findgwasloci-Tuple{DataFrames.DataFrame}"><code>GeneticsMakie.findgwasloci</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">findgwasloci(gwas::DataFrame; p::Real)
findgwasloci(gwas::Vector{DataFrame}; p::Real)</code></pre><p>Find genome-wide significant loci for <code>gwas</code> that are separated from each other by at least 1 Mb.</p><p>Alternatively, find genome-wide significant loci across multiple <code>gwas</code> that  are all separated by at least 1 Mb. <code>p</code> determines the genome-wide significance threshold,  which is 5e-8 by default.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/d595fc776b2a561097a6e79d1291ca902361d09d/src/mungesumstats.jl#L238-L248">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneticsMakie.labelgenome-Tuple{GridLayoutBase.GridPosition, AbstractString, Real, Real}" href="#GeneticsMakie.labelgenome-Tuple{GridLayoutBase.GridPosition, AbstractString, Real, Real}"><code>GeneticsMakie.labelgenome</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">labelgenome(g::GridPosition, chromosome::AbstractString, range1::Real, range2::Real)</code></pre><p>Label <code>g</code> with a given <code>chromosome</code> and genomic range between <code>range1</code> and <code>range2</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/d595fc776b2a561097a6e79d1291ca902361d09d/src/plotgenes.jl#L203-L207">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneticsMakie.mungesumstats!-Tuple{Vector{DataFrames.DataFrame}}" href="#GeneticsMakie.mungesumstats!-Tuple{Vector{DataFrames.DataFrame}}"><code>GeneticsMakie.mungesumstats!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">mungesumstats!(gwas::DataFrame)
mungesumstats!(gwas::Vector{DataFrame})</code></pre><p>Munge <code>gwas</code> by harmonizing the names of columns, their types, and P values, among others.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/d595fc776b2a561097a6e79d1291ca902361d09d/src/mungesumstats.jl#L196-L201">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneticsMakie.parsegtf!-Tuple{DataFrames.DataFrame}" href="#GeneticsMakie.parsegtf!-Tuple{DataFrames.DataFrame}"><code>GeneticsMakie.parsegtf!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">parsegtf!(gencode::DataFrame)</code></pre><p>Parse <code>gencode</code> by extracting <code>gene_id</code>, <code>gene_name</code>, <code>gene_type</code>, <code>transcript_id</code>, <code>transcript_support_level</code> information from the <code>info</code> column.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/d595fc776b2a561097a6e79d1291ca902361d09d/src/parsegtf.jl#L1-L6">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneticsMakie.plotgenes!-Tuple{Makie.Axis, AbstractString, Real, Real, DataFrames.DataFrame}" href="#GeneticsMakie.plotgenes!-Tuple{Makie.Axis, AbstractString, Real, Real, DataFrames.DataFrame}"><code>GeneticsMakie.plotgenes!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">plotgenes!(ax::Axis, chromosome::AbstractString, range1::Real, range2::Real, gencode::DataFrame)
plotgenes!(ax::Axis, chromosome::AbstractString, bp::Real, gencode::DataFrame)
plotgenes!(ax::Axis, gene::AbstractString, gencode::DataFrame)</code></pre><p>Plot collapsed gene bodies for genes within a given <code>chromosome</code> and genomic range  between <code>range1</code> and <code>range2</code>.</p><p>Alternatively, plot within a given <code>chromosome</code> and a certain <code>window</code> around a  genomic coordinate <code>bp</code> or plot within a certain <code>window</code> around <code>gene</code>.</p><p><strong>Keyword arguments</strong></p><pre><code class="nohighlight hljs">height      height of exons; default 0.25
genecolor   color of genes; default :royalblue
textcolor   color of gene labels; default :black
window      window around bp or gene; default 1e6</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/d595fc776b2a561097a6e79d1291ca902361d09d/src/plotgenes.jl#L64-L82">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneticsMakie.plotgenes!-Tuple{Makie.Axis, AbstractString, Real, Real, Tuple{AbstractVector, AbstractVector}, DataFrames.DataFrame}" href="#GeneticsMakie.plotgenes!-Tuple{Makie.Axis, AbstractString, Real, Real, Tuple{AbstractVector, AbstractVector}, DataFrames.DataFrame}"><code>GeneticsMakie.plotgenes!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">plotgenes!(ax::Axis, chromosome::AbstractString, range1::Real, range2::Real, highlight::Tuple{AbstractVector, AbstractVector}, gencode::DataFrame; height::Real)
plotgenes!(ax::Axis, chromosome::AbstractString, bp::Real, highlight::Tuple{AbstractVector, AbstractVector}, gencode::DataFrame; window::Real, height::Real)
plotgenes!(ax::Axis, gene::AbstractString, highlight::Tuple{AbstractVector, AbstractVector}, gencode::DataFrame; window::Real, height::Real)</code></pre><p>Plot gene bodies with a vector of genes highlighted by a vector of colors via <code>highlight</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/d595fc776b2a561097a6e79d1291ca902361d09d/src/plotgenes.jl#L130-L136">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneticsMakie.plotgwas!-Tuple{Makie.Axis, DataFrames.DataFrame}" href="#GeneticsMakie.plotgwas!-Tuple{Makie.Axis, DataFrames.DataFrame}"><code>GeneticsMakie.plotgwas!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">plotgwas!(ax::Axis, gwas::DataFrame)</code></pre><p>Plot <code>gwas</code> results as a Manhattan plot.</p><p><strong>Keyword arguments</strong></p><pre><code class="nohighlight hljs">ymax            maximum value for y-axis
p               genome-wide significance threshold; default 5e-8
linecolor       color of genome-wide significance line, which can be turned off by setting to nothing; default :red2
scattercolor    color of genome-wide significant variants, which can be turned off by setting to nothing; default &quot;#4DB069&quot;
chromcolors     colors of even and odd chromosomes; default [&quot;#0D0D66&quot;, &quot;#7592C8&quot;]
build::Int      human genome build; default 37 </code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/d595fc776b2a561097a6e79d1291ca902361d09d/src/plotgwas.jl#L1-L16">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneticsMakie.plotisoforms!-Tuple{Makie.Axis, AbstractString, DataFrames.DataFrame}" href="#GeneticsMakie.plotisoforms!-Tuple{Makie.Axis, AbstractString, DataFrames.DataFrame}"><code>GeneticsMakie.plotisoforms!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">plotisoforms!(ax::Axis, gene::AbstractString, gencode::DataFrame)</code></pre><p>Plot each isoform of a given <code>gene</code> on a separate row.</p><p><strong>Keyword arguments</strong></p><pre><code class="nohighlight hljs">orderby::Union{Nothing, AbstractVector{&lt;:AbstractString}}           order of isoforms; default nothing
highlight::Union{Nothing, Tuple{AbstractVector, AbstractVector}}    isoforms to be highlighted and their colors; default nothing
height                                                              height of exons; default 0.25
isoformcolor                                                        color of isoforms; default :royalblue
textcolor                                                           color of isoform labels; default :black
text::Union{Bool, Symbol}                                           position of isoform labels; default :top</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/d595fc776b2a561097a6e79d1291ca902361d09d/src/plotisoforms.jl#L66-L80">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneticsMakie.plotld-Tuple" href="#GeneticsMakie.plotld-Tuple"><code>GeneticsMakie.plotld</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">plotld(LD::AbstractMatrix; kwargs)
plotld!(ax::Axis, LD::AbstractMatrix; kwargs)</code></pre><p>Heatmap of symmetric correlation matrix <code>LD</code> with the diagonal elements on the x-axis.</p><p><strong>Keyword arguments</strong></p><ul><li><code>threshold</code>   : threshold below which values are ignored (default to <code>1/9</code>)</li><li><code>colormap</code>    : colormap of values (default to <code>cgrad(:Blues_9, 9, categorical = true)</code>)</li><li><code>colorrange</code>  : start and end points of <code>colormap</code> (default to <code>(0, 1)</code>)</li><li><code>strokewidth</code> : width of outline around heatmap boxes (default to <code>0</code>)</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/d595fc776b2a561097a6e79d1291ca902361d09d/src/plotld.jl#L1-L12">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneticsMakie.plotlocus!-Tuple{Makie.Axis, AbstractString, Real, Real, DataFrames.DataFrame}" href="#GeneticsMakie.plotlocus!-Tuple{Makie.Axis, AbstractString, Real, Real, DataFrames.DataFrame}"><code>GeneticsMakie.plotlocus!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">plotlocus!(ax::Axis, chromosome::AbstractString, range1::Real, range2::Real, gwas::DataFrame; kwargs)
plotlocus!(ax::Axis, chromosome::AbstractString, bp::Real, gwas::DataFrame; kwargs)
plotlocus!(ax::Axis, gene::AbstractString, gwas::DataFrame, gencode::DataFrame; kwargs)</code></pre><p>Plot <code>gwas</code> results within a given <code>chromosome</code> and genomic range between <code>range1</code>  and <code>range2</code>.</p><p>Alternatively, plot within a given <code>chromosome</code> and a certain <code>window</code> around a  genomic coordinate <code>bp</code> or plot within a certain <code>window</code> around <code>gene</code>.</p><p><strong>Arguments</strong></p><ul><li><code>ld::Union{Nothing, SnpData, Tuple{SnpData, Union{AbstractString, Tuple{AbstractString, Int}}}} = nothing</code>:    the reference panel for which LD is calculated.</li><li><code>ymax::Real</code>: the maximum value for y axis. </li><li><code>window::Real = 1e6</code>: the window around <code>bp</code> or <code>gene</code>. </li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/d595fc776b2a561097a6e79d1291ca902361d09d/src/plotlocus.jl#L60-L76">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneticsMakie.plotloops!-Tuple{Makie.Axis, AbstractString, Real, Real, DataFrames.AbstractDataFrame}" href="#GeneticsMakie.plotloops!-Tuple{Makie.Axis, AbstractString, Real, Real, DataFrames.AbstractDataFrame}"><code>GeneticsMakie.plotloops!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">plotloops!(ax::Axis, chromosome::AbstractString, range1::Real, range2::Real, loopdf::DataFrame)
plotloops!(ax::Axis, chromosome::AbstractString, bp::Real, loopdf::DataFrame)
plotloops!(ax::Axis, gene::AbstractString, loopdf::DataFrame, gencode::DataFrame)</code></pre><p>Plot loops present in <code>loopdf</code> within a given <code>chromosome</code> and genomic range between <code>range1</code>  and <code>range2</code>.</p><p>Alternatively, plot within a given <code>chromosome</code> and a certain <code>window</code> around a  genomic coordinate <code>bp</code> or plot within a certain <code>window</code> around <code>gene</code>.</p><p><strong>Keyword arguments</strong></p><pre><code class="nohighlight hljs">ymax            maximum value for y-axis; default 102
linewidth       line width of the arcs; default 0.25
colorarc        color of arcs; default #9658B2
colorend        color of arcs&#39; ends; default (&quot;#FFBB00&quot;, 0.5)
resolution      plot `resolution` points along x-axis within the given range; default 1000
window          window around genomic coordinate or gene; default 1e6</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/d595fc776b2a561097a6e79d1291ca902361d09d/src/plotloops.jl#L36-L56">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneticsMakie.plotqq!-Tuple{Makie.Axis, AbstractVector}" href="#GeneticsMakie.plotqq!-Tuple{Makie.Axis, AbstractVector}"><code>GeneticsMakie.plotqq!</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">plotqq!(ax::Axis, P::AbstractVector)
plotqq!(ax::Axis, gwas::DataFrame)</code></pre><p>Plot QQ plot of <code>P</code> values where the expected distribution is the uniform distribution.</p><p><strong>Keyword arguments</strong></p><pre><code class="nohighlight hljs">xstep       x-axis ticks step size; default 1
ystep       y-axis ticks step size: default 2</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/d595fc776b2a561097a6e79d1291ca902361d09d/src/plotqq.jl#L1-L12">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GeneticsMakie.plotrg-Tuple" href="#GeneticsMakie.plotrg-Tuple"><code>GeneticsMakie.plotrg</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">plotrg(r::AbstractMatrix)
plotrg!(ax::Axis, r::AbstractMatrix)</code></pre><p>Correlation plot of matrix <code>r</code>.</p><p><strong>Keyword arguments</strong></p><pre><code class="nohighlight hljs">circle          whether to draw cicles instead of rectangles; default true
diagonal        whether to visualize diagonal elements; default false
colormap        colormap of values; default :RdBu_10
colorrange      start and end points of colormap; default (-1, 1)
strokewidth     width of outline around surrounding boxes; default 0.5</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/d595fc776b2a561097a6e79d1291ca902361d09d/src/plotrg.jl#L1-L15">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../examples/peaks/">« Plotting peaks</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.4.1 on <span class="colophon-date" title="Thursday 13 June 2024 07:27">Thursday 13 June 2024</span>. Using Julia version 1.10.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
