<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Parsing GENCODE · GeneticsMakie.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link rel="canonical" href="https://mmkim1210.github.io/GeneticsMakie.jl/examples/gtf/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.044/juliamono.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.svg" alt="GeneticsMakie.jl logo"/></a><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Examples</span><ul><li class="is-active"><a class="tocitem" href>Parsing GENCODE</a></li><li><a class="tocitem" href="../summary/">Munging summary statistics</a></li><li><a class="tocitem" href="../genes/">Plotting genes</a></li><li><a class="tocitem" href="../isoforms/">Plotting isoforms</a></li><li><a class="tocitem" href="../locus/">Plotting LocusZoom</a></li><li><a class="tocitem" href="../gwas/">Plotting GWAS</a></li><li><a class="tocitem" href="../twas/">Plotting TWAS</a></li></ul></li><li><a class="tocitem" href="../../api/">API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Examples</a></li><li class="is-active"><a href>Parsing GENCODE</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Parsing GENCODE</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/master/docs/src/examples/gtf.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Parsing-GENCODE"><a class="docs-heading-anchor" href="#Parsing-GENCODE">Parsing GENCODE</a><a id="Parsing-GENCODE-1"></a><a class="docs-heading-anchor-permalink" href="#Parsing-GENCODE" title="Permalink"></a></h1><p>Install the relevant packages in the usual way.</p><pre><code class="language-julia hljs">using Pkg
Pkg.add([&quot;GeneticsMakie&quot;, &quot;CSV&quot;, &quot;DataFrames&quot;])</code></pre><p>To plot genes and isoforms, we need a transcriptome annotation. We can use  the latest <a href="https://www.gencodegenes.org/human/">GENCODE</a> annotation for  the human genome (GRCh37), where we download the comprehensive  gene annotation file in <a href="https://www.gencodegenes.org/pages/data_format.html">GTF format</a>.</p><pre><code class="language-julia hljs">using GeneticsMakie, CSV, DataFrames, Downloads
isdir(&quot;data&quot;) || mkdir(&quot;data&quot;)
url = &quot;https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_39/GRCh37_mapping/gencode.v39lift37.annotation.gtf.gz&quot;
file = basename(url)
isdir(&quot;data/gencode&quot;) || mkdir(&quot;data/gencode&quot;)
isfile(&quot;data/gencode/$(file)&quot;) || Downloads.download(url, &quot;data/gencode/$(file)&quot;)
h = [&quot;seqnames&quot;, &quot;source&quot;, &quot;feature&quot;, &quot;start&quot;, &quot;end&quot;, &quot;score&quot;, &quot;strand&quot;, &quot;phase&quot;, &quot;info&quot;]
gencode = CSV.read(&quot;data/gencode/$(file)&quot;, DataFrame; delim = &quot;\t&quot;, comment = &quot;#&quot;, header = h)</code></pre><div class="admonition is-warning"><header class="admonition-header">Human genome build</header><div class="admonition-body"><p>The latest human genome assembly is GRCh38, but we use an annotation with coordinates  from the older version (GRCh37), because a lot of the GWAS results are shared in  GRCh37 genomic coordinates. Make sure to use the matching human genome build when visualizing your results. </p></div></div><p>The ninth column of a <a href="https://uswest.ensembl.org/info/website/upload/gff.html">GTF file</a>  contains rich information about features, so we can parse this column.</p><pre><code class="language-julia hljs">GeneticsMakie.parsegtf!(gencode)</code></pre><div class="admonition is-info"><header class="admonition-header">Chromosome names</header><div class="admonition-body"><p>Chromosome names are munged to not contain “chr” prefix, and their type is <code>String</code>, since there could be non-numerical chromosome names, such as sex chromosomes and mitochondrial genome.</p></div></div><p>To reduce memory intake, we can also subset <code>gencode</code> to most commonly used columns in downstream analyses.</p><pre><code class="language-julia hljs">select!(gencode, :seqnames, :feature, :start, :end, :strand, :gene_name, :gene_type, :transcript_id)</code></pre><p>Other transcriptome annotations, such as one from RefSeq, can be used for plotting functions  as long as they contain the above columns with the right column names.</p><p>Once <code>gencode</code> is ready, we can look up where the gene is on the human genome.</p><pre><code class="language-julia hljs">GeneticsMakie.findgene(&quot;RBFOX1&quot;, gencode)</code></pre><div class="admonition is-success"><header class="admonition-header">Gene names</header><div class="admonition-body"><p>Make sure to use the correct gene name in case the gene cannot be found. The latest gene names can be looked up in databases such as <a href="https://www.genecards.org/">GeneCards</a>.</p></div></div></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../">« Home</a><a class="docs-footer-nextpage" href="../summary/">Munging summary statistics »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.16 on <span class="colophon-date" title="Sunday 1 May 2022 21:27">Sunday 1 May 2022</span>. Using Julia version 1.7.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
