<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Plotting genes · GeneticsMakie.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link rel="canonical" href="https://mmkim1210.github.io/GeneticsMakie.jl/examples/genes/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.svg" alt="GeneticsMakie.jl logo"/></a><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../gtf/">Parsing GENCODE</a></li><li><a class="tocitem" href="../summary/">Munging summary statistics</a></li><li class="is-active"><a class="tocitem" href>Plotting genes</a></li><li><a class="tocitem" href="../isoforms/">Plotting isoforms</a></li><li><a class="tocitem" href="../locus/">Plotting LocusZoom</a></li><li><a class="tocitem" href="../gwas/">Plotting GWAS</a></li><li><a class="tocitem" href="../twas/">Plotting TWAS</a></li><li><a class="tocitem" href="../loops/">Plotting loops</a></li><li><a class="tocitem" href="../peaks/">Plotting peaks</a></li></ul></li><li><a class="tocitem" href="../../api/">API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Examples</a></li><li class="is-active"><a href>Plotting genes</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Plotting genes</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/master/docs/src/examples/genes.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Plotting-genes"><a class="docs-heading-anchor" href="#Plotting-genes">Plotting genes</a><a id="Plotting-genes-1"></a><a class="docs-heading-anchor-permalink" href="#Plotting-genes" title="Permalink"></a></h1><p>After <a href="../gtf/#Parsing-GENCODE">Parsing GENCODE</a>, we can start plotting gene bodies.  <a href="https://github.com/mmkim1210/GeneticsMakie.jl"><strong>GeneticsMakie.jl</strong></a> is transparent  in that it shows all genes within a genomic window.</p><pre><code class="language-julia hljs">using Pkg
Pkg.add([&quot;GeneticsMakie&quot;, &quot;CairoMakie&quot;, &quot;DataFrames&quot;, &quot;Arrow&quot;])

using GeneticsMakie, CairoMakie, DataFrames, Arrow
url = &quot;https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_39/GRCh37_mapping/gencode.v39lift37.annotation.gtf.gz&quot;
gencode = Arrow.Table(&quot;data/gencode/$(splitext(basename(url))[1]).arrow&quot;)|&gt; DataFrame</code></pre><p>We can focus on <em>CACNA1G</em> gene as an example. We can use <code>GeneticsMakie.plotgenes!</code>  to plot all genes within a given chromosome and ± 100 Kb window around gene start and stop sites.  We can then use <code>GeneticsMakie.labelgenome</code> to label the genomic range.</p><pre><code class="language-julia hljs">isdir(&quot;figs&quot;) || mkdir(&quot;figs&quot;)
set_theme!(font = &quot;Arial&quot;)

gene = &quot;CACNA1G&quot;
chr, start, stop = GeneticsMakie.findgene(gene, gencode)
f = Figure(resolution = (306, 792))
ax = Axis(f[1, 1])
ranges = [start - 1e5, stop + 1e5]
rs = GeneticsMakie.plotgenes!(ax, chr, ranges[1], ranges[2], gencode)
GeneticsMakie.labelgenome(f[1, 1, Bottom()], chr, ranges[1], ranges[2])
rowsize!(f.layout, 1, rs)
resize_to_layout!(f)
f</code></pre><p><img src="../../figs/CACNA1G-gene.png" alt/></p><p>We can adjust the height of exons using the <code>height</code> keyword argument.</p><pre><code class="language-julia hljs">f = Figure(resolution = (306, 792))
ax = Axis(f[1, 1])
rs = GeneticsMakie.plotgenes!(ax, chr, ranges[1], ranges[2], gencode; height = 0.1)
GeneticsMakie.labelgenome(f[1, 1, Bottom()], chr, ranges[1], ranges[2])
rowsize!(f.layout, 1, rs)
resize_to_layout!(f)
f</code></pre><p><img src="../../figs/CACNA1G-gene-height0.1.png" alt/></p><p>We can change the color of genes and text using the <code>genecolor</code> and <code>textcolor</code> keyword arguments, respectively.</p><pre><code class="language-julia hljs">f = Figure(resolution = (306, 792))
ax = Axis(f[1, 1])
rs = GeneticsMakie.plotgenes!(ax, chr, ranges[1], ranges[2], gencode; height = 0.1, genecolor = :mediumorchid3, textcolor = :forestgreen)
GeneticsMakie.labelgenome(f[1, 1, Bottom()], chr, ranges[1], ranges[2])
rowsize!(f.layout, 1, rs)
resize_to_layout!(f)
f</code></pre><p><img src="../../figs/CACNA1G-gene-height0.1-color.png" alt/></p><p>Alternatively, we can visualize this locus by passing <code>gene</code> as a positional argument and  <code>window</code> as a keyword argument.</p><pre><code class="language-julia hljs">f = Figure(resolution = (306, 792))
ax = Axis(f[1, 1])
rs = GeneticsMakie.plotgenes!(ax, gene, gencode; window = 1e5, height = 0.1, genecolor = :brown3)
GeneticsMakie.labelgenome(f[1, 1, Bottom()], chr, ranges[1], ranges[2])
rowsize!(f.layout, 1, rs)
resize_to_layout!(f)
f</code></pre><p><img src="../../figs/CACNA1G-gene2.png" alt/></p><p>There is no limit to the number of genes we can visualize. Below we visualize a larger  ± 2 Mb window around <code>gene</code>.</p><pre><code class="language-julia hljs">f = Figure(resolution = (306, 792))
ax = Axis(f[1, 1])
rs = GeneticsMakie.plotgenes!(ax, gene, gencode; window = 2e6, height = 0.1)
GeneticsMakie.labelgenome(f[1, 1, Bottom()], chr, start - 2e6, stop + 2e6)
rowsize!(f.layout, 1, rs)
resize_to_layout!(f)
f</code></pre><p><img src="../../figs/CACNA1G-gene3.png" alt/></p><div class="admonition is-info"><header class="admonition-header">Gene density</header><div class="admonition-body"><p>As some regions have higher gene density than the others, it would be wise (for publication purpose) to visualize a smaller genomic window for such gene-dense regions. One can instead visualize only the genes of interest (e.g. protein coding genes) by subsetting GENCODE accordingly to such features.</p></div></div><p>We can highlight a gene or sets of genes as below. This can be useful when highlighting genes  by certain properties such as those that are protein coding or those that are  <a href="https://gnomad.broadinstitute.org/">loss-of-function intolerant</a> or those that are  significant in some sort of gene-level association. </p><pre><code class="language-julia hljs">f = Figure(resolution = (306, 792))
ax = Axis(f[1, 1])
rs = GeneticsMakie.plotgenes!(ax, chr, start, (gene, :brown3), gencode; window = 1e5, height = 0.1)
GeneticsMakie.labelgenome(f[1, 1, Bottom()], chr, start - 1e5, start + 1e5)
rowsize!(f.layout, 1, rs)
resize_to_layout!(f)
f</code></pre><p><img src="../../figs/CACNA1G-gene-highlight.png" alt/></p><pre><code class="language-julia hljs">f = Figure(resolution = (306, 792))
ax = Axis(f[1, 1])
rs = GeneticsMakie.plotgenes!(ax, chr, start, ([gene, &quot;EPN3&quot;], [:brown3, :forestgreen]), gencode; window = 1e5, height = 0.1)
GeneticsMakie.labelgenome(f[1, 1, Bottom()], chr, start - 1e5, start + 1e5)
rowsize!(f.layout, 1, rs)
resize_to_layout!(f)
f</code></pre><p><img src="../../figs/CACNA1G-gene-highlights.png" alt/></p><p>Finally, we can make additional modifications on top of the figure as needed using <a href="https://makie.juliaplots.org/stable/"><strong>Makie.jl</strong></a>.</p><pre><code class="language-julia hljs">f = Figure(resolution = (306, 792))
ax = Axis(f[1, 1])
rs = GeneticsMakie.plotgenes!(ax, chr, start, gencode; window = 1e5, height = 0.1)
GeneticsMakie.labelgenome(f[1, 1, Bottom()], chr, start - 1e5, start + 1e5)
rowsize!(f.layout, 1, rs)
resize_to_layout!(f)
vlines!(ax, start, color = (:gold, 0.5), linewidth = 0.5)
vlines!(ax, stop, color = (:gold, 0.5), linewidth = 0.5)
f</code></pre><p><img src="../../figs/CACNA1G-gene-line.png" alt/></p><p>Then we can save the figure as below.</p><pre><code class="language-julia hljs">save(&quot;figs/$(gene)-gene.png&quot;, f, px_per_unit = 4)</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../summary/">« Munging summary statistics</a><a class="docs-footer-nextpage" href="../isoforms/">Plotting isoforms »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Sunday 2 October 2022 17:57">Sunday 2 October 2022</span>. Using Julia version 1.8.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
