<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Plotting LocusZoom · GeneticsMakie.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link rel="canonical" href="https://mmkim1210.github.io/GeneticsMakie.jl/examples/locus/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.svg" alt="GeneticsMakie.jl logo"/></a><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../gtf/">Parsing GENCODE</a></li><li><a class="tocitem" href="../summary/">Munging summary statistics</a></li><li><a class="tocitem" href="../genes/">Plotting genes</a></li><li><a class="tocitem" href="../isoforms/">Plotting isoforms</a></li><li class="is-active"><a class="tocitem" href>Plotting LocusZoom</a></li><li><a class="tocitem" href="../gwas/">Plotting GWAS</a></li><li><a class="tocitem" href="../twas/">Plotting TWAS</a></li></ul></li><li><a class="tocitem" href="../../api/">API</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Examples</a></li><li class="is-active"><a href>Plotting LocusZoom</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Plotting LocusZoom</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/mmkim1210/GeneticsMakie.jl/blob/master/docs/src/examples/locus.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Plotting-LocusZooom"><a class="docs-heading-anchor" href="#Plotting-LocusZooom">Plotting LocusZooom</a><a id="Plotting-LocusZooom-1"></a><a class="docs-heading-anchor-permalink" href="#Plotting-LocusZooom" title="Permalink"></a></h1><p>After <a href="../gtf/#Parsing-GENCODE">Parsing GENCODE</a> and <a href="../summary/#Munging-summary-statistics">Munging summary statistics</a>, we can now put the pieces together to draw the backbone of a LocusZoom plot.  We focus on <em>JAZF1</em> locus as an example, which reaches strong genome-wide significance in GWAS for height. By default, <code>GeneticsMakie.plotlocus!</code> returns a straightforward scatter plot.</p><pre><code class="language-julia hljs">gene = &quot;JAZF1&quot;
chr, start, stop = GeneticsMakie.findgene(gene, gencode)
range1 = start - 1e6
range2 = stop + 1e6

n = length(df)
titles = [&quot;Height (Yengo et al. 2018)&quot;, &quot;Weight (Yengo et al. 2018)&quot;]
f = Figure(resolution = (306, 792))
axs = [Axis(f[i, 1]) for i in 1:(n + 1)]
for i in 1:n
    GeneticsMakie.plotlocus!(axs[i], chr, range1, range2, df[i])
    rowsize!(f.layout, i, 30)
    lines!(axs[i], [range1, range2], fill(-log(10, 5e-8), 2), color = (:purple, 0.5), linewidth = 0.5)
    Label(f[i, 1, Top()], &quot;$(titles[i])&quot;, textsize = 6, halign = :left, padding = (7.5, 0, -5, 0))
end
rs = GeneticsMakie.plotgenes!(axs[n + 1], chr, range1, range2, gencode; height = 0.1)
rowsize!(f.layout, n + 1, rs)
GeneticsMakie.labelgenome(f[n + 1, 1, Bottom()], chr, range1, range2)
Colorbar(f[1:n, 2], limits = (0, 1), ticks = 0:1:1, height = 20,
    colormap = (:gray60, :red2), label = &quot;LD&quot;, ticksize = 0, tickwidth = 0,
    tickalign = 0, ticklabelsize = 6, flip_vertical_label = true,
    labelsize = 6, width = 5, spinewidth = 0.5)
Label(f[1:n, 0], text = &quot;-log[p]&quot;, textsize = 6, rotation = pi / 2)
rowgap!(f.layout, 5)
colgap!(f.layout, 5)
for i in 1:(n + 1)
    vlines!(axs[i], start, color = (:gold, 0.5), linewidth = 0.5)
    vlines!(axs[i], stop, color = (:gold, 0.5), linewidth = 0.5)
end
resize_to_layout!(f)
f</code></pre><p><img src="../../figs/JAZF1-locuszoom.png" alt/></p><p>To color variants by linkage disequilibrium (LD), we need a reference panel. If we already have one, we can use <a href="https://openmendel.github.io/SnpArrays.jl/latest/"><strong>SnpArrays.jl</strong></a> to read in PLINK bed files. If not, we can download one as below. For the time being,  we only download and convert a single chromosome from the 1000 Genomes Project. </p><pre><code class="language-julia hljs">using Pkg
Pkg.add(&quot;SnpArrays&quot;)</code></pre><pre><code class="language-julia hljs">using SnpArrays, Downloads
# Download 1000 Genomes data for a single chromosome
beagle = &quot;http://bochet.gcc.biostat.washington.edu/beagle/1000_Genomes_phase3_v5a&quot;
url = joinpath(beagle, &quot;b37.vcf/chr$(chr).1kg.phase3.v5a.vcf.gz&quot;)
vcf = basename(url)
isdir(&quot;data/1kg&quot;) || mkdir(&quot;data/1kg&quot;)
isfile(&quot;data/1kg/$(vcf)&quot;) || Downloads.download(url, &quot;data/1kg/$(vcf)&quot;)
# Convert vcf file to plink bed file (this step takes a while)
isfile(&quot;data/1kg/$(replace(vcf, &quot;.vcf.gz&quot; =&gt; &quot;.bed&quot;))&quot;) || vcf2plink(&quot;data/1kg/$(vcf)&quot;, &quot;data/1kg/$(replace(vcf, &quot;.vcf.gz&quot; =&gt; &quot;&quot;))&quot;)
# Download sample metadata
url = joinpath(beagle, &quot;sample_info/integrated_call_samples_v3.20130502.ALL.panel&quot;)
meta = basename(url) 
isfile(&quot;data/1kg/$(meta)&quot;) || Downloads.download(url, &quot;data/1kg/$(meta)&quot;)
# Subset data to the genomic region of interest and European samples
kgp = SnpData(&quot;data/1kg/$(replace(vcf, &quot;.vcf.gz&quot; =&gt; &quot;&quot;))&quot;)
meta = CSV.read(&quot;data/1kg/$(meta)&quot;, DataFrame)
eur = meta.sample[meta.super_pop .== &quot;EUR&quot;]
colinds = findall((kgp.snp_info.position .&gt;= range1) .&amp; (kgp.snp_info.position .&lt;= range2))
rowinds = findall(in(eur), kgp.person_info.iid)
file = replace(vcf, &quot;.vcf.gz&quot; =&gt; &quot;.eur&quot;)
SnpArrays.filter(kgp, rowinds, colinds; des = &quot;data/1kg/$(file)&quot;)
# Apply minor allele frequency &gt; 0.05 filter
kgp = SnpData(&quot;data/1kg/$(file)&quot;)
colinds = SnpArrays.filter(kgp.snparray; min_maf = 0.05)[2]
file = file * &quot;.maf0.05&quot;
SnpArrays.filter(kgp, trues(size(kgp)[1]), colinds; des = &quot;data/1kg/$(file)&quot;)
kgp = SnpData(&quot;data/1kg/$(file)&quot;)</code></pre><p>We can color variants by LD with the index/sentinel SNP by using the <code>ld</code> keyword argument.</p><pre><code class="language-julia hljs">f = Figure(resolution = (306, 792))
axs = [Axis(f[i, 1]) for i in 1:(n + 1)]
for i in 1:n
    GeneticsMakie.plotlocus!(axs[i], chr, range1, range2, df[i], ld = kgp)
    rowsize!(f.layout, i, 30)
    lines!(axs[i], [range1, range2], fill(-log(10, 5e-8), 2), color = (:purple, 0.5), linewidth = 0.5)
    Label(f[i, 1, Top()], &quot;$(titles[i])&quot;, textsize = 6, halign = :left, padding = (7.5, 0, -5, 0))
end
rs = GeneticsMakie.plotgenes!(axs[n + 1], chr, range1, range2, gencode; height = 0.1)
rowsize!(f.layout, n + 1, rs)
GeneticsMakie.labelgenome(f[n + 1, 1, Bottom()], chr, range1, range2)
Colorbar(f[1:n, 2], limits = (0, 1), ticks = 0:1:1, height = 20,
    colormap = (:gray60, :red2), label = &quot;LD&quot;, ticksize = 0, tickwidth = 0,
    tickalign = 0, ticklabelsize = 6, flip_vertical_label = true,
    labelsize = 6, width = 5, spinewidth = 0.5)
Label(f[1:n, 0], text = &quot;-log[p]&quot;, textsize = 6, rotation = pi / 2)
rowgap!(f.layout, 5)
colgap!(f.layout, 5)
for i in 1:(n + 1)
    vlines!(axs[i], start, color = (:gold, 0.5), linewidth = 0.5)
    vlines!(axs[i], stop, color = (:gold, 0.5), linewidth = 0.5)
end
resize_to_layout!(f)
f</code></pre><p><img src="../../figs/JAZF1-locuszoom-ld.png" alt/></p><p>We can also color variants by LD with the same SNP by using the <code>ld</code> keyword argument.</p><pre><code class="language-julia hljs">f = Figure(resolution = (306, 792))
axs = [Axis(f[i, 1]) for i in 1:(n + 1)]
for i in 1:n
    GeneticsMakie.plotlocus!(axs[i], chr, range1, range2, df[i], ld = (kgp, &quot;rs508347&quot;))
    rowsize!(f.layout, i, 30)
    lines!(axs[i], [range1, range2], fill(-log(10, 5e-8), 2), color = (:purple, 0.5), linewidth = 0.5)
    Label(f[i, 1, Top()], &quot;$(titles[i])&quot;, textsize = 6, halign = :left, padding = (7.5, 0, -5, 0))
end
rs = GeneticsMakie.plotgenes!(axs[n + 1], chr, range1, range2, gencode; height = 0.1)
rowsize!(f.layout, n + 1, rs)
GeneticsMakie.labelgenome(f[n + 1, 1, Bottom()], chr, range1, range2)
Colorbar(f[1:n, 2], limits = (0, 1), ticks = 0:1:1, height = 20,
    colormap = (:gray60, :red2), label = &quot;LD&quot;, ticksize = 0, tickwidth = 0,
    tickalign = 0, ticklabelsize = 6, flip_vertical_label = true,
    labelsize = 6, width = 5, spinewidth = 0.5)
Label(f[1:n, 0], text = &quot;-log[p]&quot;, textsize = 6, rotation = pi / 2)
rowgap!(f.layout, 5)
colgap!(f.layout, 5)
for i in 1:(n + 1)
    vlines!(axs[i], start, color = (:gold, 0.5), linewidth = 0.5)
    vlines!(axs[i], stop, color = (:gold, 0.5), linewidth = 0.5)
end
resize_to_layout!(f)
f</code></pre><p><img src="../../figs/JAZF1-locuszoom-ld-snp.png" alt/></p><p>By using <a href="https://makie.juliaplots.org/stable/"><strong>Makie.jl</strong></a>&#39;s layout tools,  it becomes easy to draw additional tracks. For example, in a separate track,  the variants could be colored or could have varying sizes depending on their minor allele frequency.  In another example, the variants could be colored based on their inclusion in a  credible set post-fine-mapping.</p><div class="admonition is-info"><header class="admonition-header">Plotting the intersection of SNPs, not the union</header><div class="admonition-body"><p><code>GeneticsMakie.plotlocus!</code> plots only the variants that are present in the reference panel,  when the <code>ld</code> keyword argument is specified. Although SNPs that are missing in the reference panel could be plotted differently (e.g. with varying transparency and shape), <a href="https://github.com/mmkim1210/GeneticsMakie.jl"><strong>GeneticsMakie.jl</strong></a> is designed to visualize 100s of phenotypes simultaneously in which case such discrepancy is hard to tell and  is confusing. Hence, for more direct comparison of loci across phenotypes,  only the variants that are found in the reference panel are shown.</p></div></div><div class="admonition is-warning"><header class="admonition-header">Extremely small P values</header><div class="admonition-body"><p>There are several GWAS loci that harbor extremely small P values, in which cases the P values will be clamped to the smallest floating point number. Such cases are going to be more common in phenotypes that are reaching saturation in terms of GWAS discovery (e.g. height). In those cases, it is more commonplace to observe allelic heterogneity, and it might be more appropriate to plot alternative measures of strength of association (e.g. Z scores).</p></div></div><div class="admonition is-info"><header class="admonition-header">Patterns of LD</header><div class="admonition-body"><p>Oftentimes, chunks of LD blocks hug the gene boundaries.</p></div></div><div class="admonition is-success"><header class="admonition-header">Covering the entire genome</header><div class="admonition-body"><p>Visualizing 1,500 genomic regions with 2 Mb window will more or less cover the entire human genome. Note that empirically speaking, the probability of an arbitrary 2 Mb window  harboring at least one genome-wide significant hit across multiple phenotypes is  higher than not harboring any significant association.</p></div></div></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../isoforms/">« Plotting isoforms</a><a class="docs-footer-nextpage" href="../gwas/">Plotting GWAS »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.15 on <span class="colophon-date" title="Saturday 19 March 2022 17:47">Saturday 19 March 2022</span>. Using Julia version 1.7.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
